### PR TITLE
Pipeshot: craftable makeshift shotgun shells

### PR DESCRIPTION
**Pull request by NightmareChamillian**
[input][feature][balance]
## About the PR 
This PR adds craftable shotgun shells for the slamgun to the game:
image
These can be made by cutting a filled pipe frame with wirecutters, which will yield a cut empty set of hulls. Adding glass or scrap to these will produce 4 pipeshot shells of the corrosponding type. These shells are now usable in any shotgun
Video Demos:
Raw damage:
https://www.youtube.com/watch?v=gpwi_8YTm-Q
Interaction with armor
https://www.youtube.com/watch?v=FILrLy10U0s

Shells:
image
Both shells have a maximum total damage of 24, but suffer from unreliable spread and heavy falloff. Despite outclassing the .22 bullet and phaser in terms of pure damage, they're definitely a class below the two due to the firerates and damage falloff, but allow for a higher skill ceiling.

-Glass shells fire 6 projectiles, and have heavy decay, doing 1-3 damage at a 2 tile range. This is accentuated by the 30 degree spread it suffers from. It's also worth pointing out these don't embed. However, they're far easier to get ahold of, as shards can be found practically everywhere ingame. Depending on feedback I might want to reduce the damage or number of pellets because of this, or up the required number of shards to yield the ammunition.

-Scrap shells fire only 3 projectiles, but with a much tighter (15 degree) spread, and less falloff. Despite being similar on paper, they by far trump the glass shells for the tradeoff of the user having to venture out and find material scrap. They offer decent damage at ranges of 3-5 tiles, dealing 12-18 in this range, but still have a heavy falloff that keeps them from straying too far from the approach, fire, back off and reload rhythm. Again, if the consensus is this makes it too easy to get ahold of, I can adjust the numbers or required chunks of scrap.

## Why's this needed? 

Improvised firearms are one of the most fun ways of doing damage in the game, and deepen the setting by widening the niche of handmade firearms- Nanotrasen uses lasers, the syndicate use conventional kinetics, but theives and viglantes, too poor for either, have to improvise their own means.

Furthermore, it gives the slamgun purpose. If you want to use the slamgun, you either have to break into bar, or armory, meaning it's really only ever going to be thought about by spies and hard mode traitors. I'll be frank and say the slamgun right now sucks- beyond the absurd strategy of carrying 5 of them loaded with rubber slugs, it's far worse in reliable damage and harder to find ammo for than making a zip gun or finding a phaser. It gives the slamgun a niche- an improvised firearm higher damage than a .22 loaded zip gun, but riskier and with more mastery involved due to the reload.

Lastly, and most importantly, it make the slam gun fun.. It's fun to rush in and pointblank someone with glass to see their character get pushed by the projectiles, then step back to reload or swap to melee. It's satisfying to hit two scrap projectiles on a target 3 tiles away and see the damage tick up. The small yield prevents the unsavory tactic of carrying a backpack full of loaded slamguns, and the spread makes it a force to be reckoned with in close quarters.

## Changelog

```changelog
(u)NightmarechaMillian
(*)Added craftable slamgun shells to the game.
```

### CODE REVIEW

File: code/obj/item/gun/ammo.dm [hunk 1]
```diff
@@ -675,6 +675,35 @@
 	icon_empty = "12-0"
 	sound_load = 'sound/weapons/gunload_heavy.ogg'
 
+/obj/item/ammo/bullets/pipeshot
```
Comment (blackrep) [inline code/obj/item/gun/ammo.dm:678]:
```suggestion
ABSTRACT_TYPE(/obj/item/ammo/bullets/pipeshot)
/obj/item/ammo/bullets/pipeshot
```

File: code/obj/item/gun/ammo.dm [hunk 2]
```diff
@@ -675,6 +675,35 @@
...
+	name = "pipeshot"
+	desc = "A parent item! If you see this contact a coder."
+	ammo_type = new/datum/projectile/special/spreader/buckshot_burst
+	icon_state = "makeshiftempty"
+	amount_left = 4.0
+	max_amount = 4.0
+	ammo_cat = AMMO_SHOTGUN_SCRAP
+	delete_on_reload = TRUE
+	icon_dynamic = FALSE
+	icon_empty = "makeshiftempty"
```
Comment (blackrep) [inline code/obj/item/gun/ammo.dm:689]: If `delete_on_reload = TRUE` is this `icon_empty` ever used?

File: code/obj/item/gun/ammo.dm [hunk 3]
```diff
@@ -675,6 +675,35 @@
...
+	sname = "pipeshot"
+	name = "pipeshot"
+	desc = "A parent item! If you see this contact a coder."
+	ammo_type = new/datum/projectile/special/spreader/buckshot_burst
+	icon_state = "makeshiftempty"
+	amount_left = 4.0
+	max_amount = 4.0
+	ammo_cat = AMMO_SHOTGUN_SCRAP
+	delete_on_reload = TRUE
+	icon_dynamic = FALSE
```
Comment (blackrep) [inline code/obj/item/gun/ammo.dm:688]: Already `FALSE` in parent `/obj/item/ammo/bullets` no need to redefine it again.
```suggestion
```

File: _std/defines/projectiles.dm [hunk 4]
```diff
@@ -32,9 +32,11 @@
 
 
 //ammo categories
+#define AMMO_SHOTGUN_SCRAP "shotgun_scrap"
```
Comment (blackrep) [inline _std/defines/projectiles.dm:35]: Personal take, I would like to see it usable in all shotguns. Non-antags can obtain riot shotguns out in the debris field & azones, why force them to only use the slam gun / coach gun if they're able to obtain something better?

Not an issue for antags considering they can already break into the armory or purchase better ammo from uplink / syndicate listening post.

File: code/obj/item/grenades.dm [hunk 5]
```diff
@@ -1458,6 +1460,12 @@ ABSTRACT_TYPE(/obj/item/old_grenade/spawner)
 				name = "hollow pipe frame"
 			src.flags |= NOSPLASH
 
+		if(issnippingtool(W) && state == 3) //pipeshot crafting
```
Comment (blackrep) [inline code/obj/item/grenades.dm:1463]: Correct me if I'm wrong, but this appears like it never considers what chem is used as the propellent.

Example:
Pipebombs use the volatility / strength var to determine if a pipebomb will explode or crumble. Your shotgun shells will accept water is a valid propellent and act the same as black powder.

Two suggestions:
1. Re-use the strength check pipebomb code, if the strength is below a certain threshold (e.x they just used water) give them dud rounds.
2. Make the strength of the propellent matter, increase / decrease speed of projectile. Or maybe even give it a damage boost.

File: code/obj/item/gun/ammo.dm (hunk 2 above)
Comment (NightmareChamillian) [reply code/obj/item/gun/ammo.dm:689]: nope, will remove

File: code/obj/item/grenades.dm (hunk 5 above)
Comment (NightmareChamillian) [reply code/obj/item/grenades.dm:1463]: They now won't craft if you don't use an explosive chem

File: code/modules/projectiles/special.dm [hunk 6]
```diff
@@ -212,6 +212,31 @@ ABSTRACT_TYPE(/datum/projectile/special)
...
+
+/datum/projectile/special/spreader/buckshot_burst/scrap
+	name = "fragments"
+	sname = "fragments"
+	cost = 1
+	pellets_to_fire = 3
+	spread_projectile_type = /datum/projectile/bullet/improvscrap
+	shot_sound = 'sound/weapons/shotgunshot.ogg'
+	speed_max = 20
+	speed_min = 60
```
Comment (Tarmunora) [inline code/modules/projectiles/special.dm:235]: 2 things:

- You have a higher min than max
- too much variance here tends to feel really janky - I've found that a min of 28 and a max of 36 or so is about as far as I could go without it looking weird.

Ditto above for glass projectile

File: code/obj/item/gun/ammo.dm [hunk 7]
```diff
@@ -675,6 +675,35 @@
...
+ABSTRACT_TYPE(/obj/item/ammo/bullets/pipeshot)
+/obj/item/ammo/bullets/pipeshot
+	sname = "pipeshot"
+	name = "pipeshot"
+	desc = "A parent item! If you see this contact a coder."
+	ammo_type = new/datum/projectile/special/spreader/buckshot_burst
+	icon_state = "makeshiftempty"
+	amount_left = 4.0
+	max_amount = 4.0
+	ammo_cat = AMMO_SHOTGUN_LOW
```
Comment (Tarmunora) [inline code/obj/item/gun/ammo.dm:687]:
```suggestion
	ammo_cat = AMMO_SHOTGUN_HIGH
```
don't really want these loaded in flareguns - would feel weird

**Review by Tarmunora — CHANGES_REQUESTED** [review]

**Pull request merged**